### DISCUSSION TITLE
Support 3rd party hosted AMP mirror

### DISCUSSION THREAD

**Discussion by amerzad** [vercel/next.js#23960] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

Yes, the problem is that I needed to stream line the development of a content rich website by stopping duplicating code/design for AMP and non-AMP pages, especially after AMP-Script was publicly supported, and at the same time I needed a way to be future-proof for the Web Vitals imposed by latest SEO best practices, so I wanted to do an SSG AMP only website using Next.js, but I discovered that some major countries are blocking the official AMP cdn for some reason, so that blocked the whole idea.

## Describe the solution you'd like

I think the best way is to use AMP only (`amp=true`) and allow some kind of runtime configuration for providing the base url for the 3rd party AMP library mirror when the user agent is not a bot, or maybe on any custom condition, and on SSG we should have two page caches, one for bot requests and one for non-bot requests, or even for simplicity, there could be one cache with a function which will replace the official AMP cdn code with the 3rd party mirror code for non-bot requests.

Maybe it could be a general feature where some code will be provided per request to determine which cache among multiple caches will be used, and another code or configuration will be used to determine the different caches needed at build time, and the SSG builder will send a request for every cache case with some parameter to the page when doing the SSG so the page will return the right content per props/cache case.

## Describe alternatives you've considered

I came up with the idea that I can use Hybrid mode in Next.js and make all the pages with `unstable_runtimeJS = false`, and have my own self-hosted mirror of the AMP library on Github, and inject all the amp scripts I need from this mirror in the head of every page when `useAmp` returns false.

This worked fine, and implementing all the interactive features in AMP in that multi-user website was simpler than what we expected, but I faced two problems:
- The some what immature feature support for Hybrid AMP/SSG which manifested in this bug report: #17193 (and several other issues)
- The developer experience in Next.js is not nice, especially with the so many development time errors in the console because `unstable_runtimeJS` is not meant to be developed fully as mentioned in the PR of the developer

## Additional context

I think there is some talks about that Google cache will eventually support 3rd party mirrors of the AMP library and provide some way to validate these pages, but they didn't put any time limit and it seems to be not that simple, and if happened, it may get blocked by these countries some way, so hosting my own mirror of the library may stay the only option for my case.

**Comment by c0b41** [comment]
Well you can be already do this
download amp assets public directory

```
npx @ampproject/toolbox-cli download --dest ./public/amp --rtv 012004030010070
```
put next config for ampUrlPrefix

```
experimental: {
     amp: {
      optimizer: {
        ampRuntimeVersion: "012004030010070",
        ampUrlPrefix: "/amp",
        verbose: true,
        blurredPlaceholders: false,
      }
    }
 }
```

you can bypass country block.

**Comment by amerzad** [comment]
Thanks @c0b41 for your reply.

I actually know about this option, but according to my understanding, this will cause the page to be invalidated for Google Cache, which is not acceptable, am I not right?

I need to have two versions of my pages, one for Google Cache with the official AMP CDN, and one for normal users with my own AMP mirror/CDN, or even in the same project.

**Comment by c0b41** [comment]
@amerzad I don't know :/ maybe @sebastianbenz can answer this
